aggregators/aggregator.go

```diff
@@ -147,20 +149,28 @@ func (a *Aggregator) AggregateBatch(
 			bytesIn, err := a.aggregateAPMEvent(ctx, cmk, e)
 			if err != nil {
 				errs = append(errs, err)
+				failBytes += int64(bytesIn)
+			} else {
+				successBytes += int64(bytesIn)
 			}
-			totalBytesIn += int64(bytesIn)
 		}
 		a.cachedEvents.add(ivl, id, float64(len(*b)))
 	}
 
-	cmIDAttrSet := attribute.NewSet(cmIDAttrs...)
-	a.metrics.RequestsTotal.Add(ctx, 1, metric.WithAttributeSet(cmIDAttrSet))
-	a.metrics.BytesIngested.Add(ctx, totalBytesIn, metric.WithAttributeSet(cmIDAttrSet))
+	var (
+		err     error
+		attrSet attribute.Set
+	)
 	if len(errs) > 0 {
-		a.metrics.RequestsFailed.Add(ctx, 1, metric.WithAttributeSet(cmIDAttrSet))
-		return fmt.Errorf("failed batch aggregation:\n%w", errors.Join(errs...))
-	}
-	return nil
+		attrSet = attribute.NewSet(append(cmIDAttrs, telemetry.WithFailure())...)
+		a.metrics.BytesProcessed.Add(ctx, failBytes, metric.WithAttributeSet(attrSet))
+		err = fmt.Errorf("failed batch aggregation:\n%w", errors.Join(errs...))
+	} else {
+		attrSet = attribute.NewSet(append(cmIDAttrs, telemetry.WithSuccess())...)
+	}
+	a.metrics.BytesProcessed.Add(ctx, successBytes, metric.WithAttributeSet(attrSet))
+	a.metrics.RequestsCount.Add(ctx, 1, metric.WithAttributeSet(attrSet))
```

Comment:
This means that attrSet has both `outcome: failure` and `outcome: success` attributes. I suppose the last one wins, but is this intentional? Should we be updating BytesProcessed and RequestsCount twice? Once for failure, once for success?

Reply:
> I suppose the last one wins, but is this intentional? 
> Should we be updating BytesProcessed and RequestsCount twice? Once for failure, once for success?

Yes, it is intentional. Since each event is processed for all configured aggregation intervals, it is possible that we have a few failures and a few successes. The expectation is that the failure is added when there is a failure and success is added every time. The success outcome will be recorded `0` if all fails. Do you see some issue in this approach? The attribute sets are different w.r.t. `outcome` so they should be recorded individually and I don't expect any conflicts.

Reply:
Actually I think I misread it before. With the new code, similar to before:

```go
	if len(errs) > 0 {
		attrSetOpt = metric.WithAttributeSet(
			attribute.NewSet(append(cmIDAttrs, telemetry.WithFailure())...),
		)
		a.metrics.BytesProcessed.Add(ctx, failBytes, attrSetOpt)
		err = fmt.Errorf("failed batch aggregation:\n%w", errors.Join(errs...))
	} else {
		attrSetOpt = metric.WithAttributeSet(
			attribute.NewSet(append(cmIDAttrs, telemetry.WithSuccess())...),
		)
	}
	a.metrics.BytesProcessed.Add(ctx, successBytes, attrSetOpt)
```

In the case where some errors occurred, doesn't this mean we're
- adding `failedBytes` to `BytesProcessed` with `outcome: failure`
- adding `successBytes` to `BytesProcessed` also with `outcome: failure` (because we updated attrSetOpt above)

That second one doesn't seem right?

Do we want this instead?

```go
	if len(errs) > 0 {
		a.metrics.BytesProcessed.Add(ctx, failBytes, metric.WithAttributeSet(
			attribute.NewSet(append(cmIDAttrs, telemetry.WithFailure())...),
		)
		err = fmt.Errorf("failed batch aggregation:\n%w", errors.Join(errs...))
	}
	a.metrics.BytesProcessed.Add(ctx, successBytes, metric.WithAttributeSet(
		attribute.NewSet(append(cmIDAttrs, telemetry.WithSuccess())...),
	))
```

Reply:
You are correct 🤦‍♂️.

Reply:
> Thanks for catching this blunder. I tried writing a test for it but failing AggregateBatch in test is quite difficult (let me know if you have some ideas here). For now I have fixed the code.

Yeah I'm not sure how we can force that to fail. Maybe that's a good thing? :smile: